lib/services/watcher.go

```diff
@@ -1666,3 +1666,159 @@ func (p *accessRequestCollector) processEventAndUpdateCurrent(ctx context.Contex
 }
 
 func (*accessRequestCollector) notifyStale() {}
+
+// SAMLIdPServiceProviderWatcherConfig is a SAMLIdPServiceProviderWatcher configuration.
+type SAMLIdPServiceProviderWatcherConfig struct {
+	// ResourceWatcherConfig is the resource watcher configuration.
+	ResourceWatcherConfig
+	// SAMLIdPServiceProviders is responsible for fetching access SAML IdP service providers.
+	SAMLIdPServiceProviders
+	// PageSize is the number of providers to list at a time.
+	PageSize int
+	// SAMLIdPServiceProvidersC receives up-to-date list of all SAML IdP service provider resources.
+	SAMLIdPServiceProvidersC chan types.SAMLIdPServiceProviders
+}
+
+// CheckAndSetDefaults checks parameters and sets default values.
+func (cfg *SAMLIdPServiceProviderWatcherConfig) CheckAndSetDefaults() error {
+	if err := cfg.ResourceWatcherConfig.CheckAndSetDefaults(); err != nil {
+		return trace.Wrap(err)
+	}
+	if cfg.SAMLIdPServiceProviders == nil {
+		getter, ok := cfg.Client.(SAMLIdPServiceProviders)
+		if !ok {
+			return trace.BadParameter("missing parameter SAMLIdPServiceProviders and Client not usable as SAMLIdPServiceProviders")
+		}
+		cfg.SAMLIdPServiceProviders = getter
+	}
+	if cfg.PageSize == 0 {
+		cfg.PageSize = 200
+	}
+	if cfg.SAMLIdPServiceProvidersC == nil {
+		cfg.SAMLIdPServiceProvidersC = make(chan types.SAMLIdPServiceProviders)
+	}
+	return nil
+}
+
+// NewSAMLIdPServiceProviderWatcher returns a new instance of SAMLIdPServiceProviderWatcher.
+func NewSAMLIdPServiceProviderWatcher(ctx context.Context, cfg SAMLIdPServiceProviderWatcherConfig) (*SAMLIdPServiceProviderWatcher, error) {
+	if err := cfg.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	collector := &samlIDPServiceProviderCollector{
+		SAMLIdPServiceProviderWatcherConfig: cfg,
+		initializationC:                     make(chan struct{}),
+	}
+	watcher, err := newResourceWatcher(ctx, collector, cfg.ResourceWatcherConfig)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return &SAMLIdPServiceProviderWatcher{watcher, collector}, nil
+}
+
+// SAMLIdPServiceProviderWatcher is built on top of resourceWatcher to monitor SAML IdP service provider resources.
+type SAMLIdPServiceProviderWatcher struct {
+	*resourceWatcher
+	*samlIDPServiceProviderCollector
+}
+
+// samlIDPServiceProviderCollector accompanies resourceWatcher when monitoring SAML IdP service provider resources.
+type samlIDPServiceProviderCollector struct {
+	// SAMLIdPServiceProviderWatcherConfig is the watcher configuration.
+	SAMLIdPServiceProviderWatcherConfig
+	// current holds a map of the currently known SAML IdP service provider resources.
+	current map[string]types.SAMLIdPServiceProvider
+	// lock protects the "current" map.
+	lock sync.RWMutex
```

Comment:
Suggestion: a somewhat-common pattern is to call the lock `mu` and define it before the fields it guards. (Feel free to disregard.)

```suggestion
	// mu guards `current`.
	mu sync.RWMutex
	// current holds a map of (currently known) provider name to resource.
	current map[string]types.SAMLIdPServiceProvider
```